docs/source-2.0/guides/codegen/mapping-shapes-to-languages.rst

```diff
@@ -0,0 +1,629 @@
+maps, etc. Some target environments allow types to be created using a
+kind of literal syntax that does not perform any custom initialization.
+In these cases, it may be necessary to use a constructor method in order
+to set members to their default zero values if needed.
+
+Structure and union members are ordered based on the order they are
+defined in the model. When adding new members, they should be added to
+the end of the structure. While this allows code generators like C++ to
+maintain ABI compatibility, it requires extreme levels of rigor to
+enforce that every change will be ABI compatible.
+
+
+Error structures
+~~~~~~~~~~~~~~~~
+
+Structures marked with the ``@error`` trait should be code generated as
+a kind of error type or exception type in the target environment. A good
+design goal for errors generated from Smithy models is to allow generic
+abstractions to work across generated Smithy clients. For example,
+developers should be able to create a middleware that can be used with
+any Smithy generated client to check if an error is a client error,
+server error, retryable, or throttling error. The :ref:`retryable-trait`
+is used to describe if an error can be retried, and the ``throttling``
+property of this trait describes if the error is due to throttling. This
+information should be exposed by the generated type in some way.
+
+Errors could have a kind of hierarchy resembling the following (note
+that other error conditions like networking errors need to be accounted
+for as well):
+
+- Service specific error: a top-level error type generated specifically
+  for every error the service can return. This error is used when an
+  unmodeled exception is encountered.
+- Client Error: Error used when an ``@error`` trait is set to "client".
+- Server Error: Error used when an ``@error`` trait is set to "server".
+
+
+Union
+-----
+
+The union type represents a `tagged union data
+structure <https://en.wikipedia.org/wiki/Tagged_union>`__ that can take
+on several different, but fixed, types. Unions function similarly to
+structures except that only one member can be used at any one time.
+Unions are always code generated and use the name provided in the model.
+Code generators should provide some kind of abstraction to make union
+types easier to use. For example, if a target environment supports sum
+types or discriminated unions, use them. Sealed classes with specific
+subtypes for each variant of the union are also good options.
+
+- The member that is set in a union cannot be optional.
+- There must be exactly one member of the union set to a non-null
+  value.
+- Clients must account for unknown union values by storing the name of
+  the unknown variant.
+
+
+Unit types in unions
+~~~~~~~~~~~~~~~~~~~~
+
+Union members may target Smithy's built-in unit type,
+``smithy.api#Unit``, meaning the variant has no meaningful value. If a
+unit type member targets the unit type, implementations should generate
```

Comment:
Should the first "unit type" on this line be removed?

Reply:
Yeah will modify